src/query/executor/types.go

```diff
@@ -0,0 +1,88 @@
+// Copyright (c) 2019 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package executor
+
+import (
+	"context"
+	"time"
+
+	qcost "github.com/m3db/m3/src/query/cost"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/parser"
+	"github.com/m3db/m3/src/query/storage"
+
+	"github.com/uber-go/tally"
+)
+
+// Engine executes a Query.
+type Engine interface {
+	// Execute runs the query and closes the results channel once done.
+	Execute(
+		ctx context.Context,
+		query *storage.FetchQuery,
+		results chan *storage.QueryResult,
+	)
+
+	// ExecuteExpr runs the query DAG and closes the results channel once done.
+	// nolint: unparam
+	ExecuteExpr(
+		ctx context.Context,
+		parser parser.Parser,
+		params models.RequestParams,
+		results chan Query,
+	)
+
+	// Close kills all running queries and prevents new queries from being attached.
+	Close() error
+
+	// Opts returns the engine options.
+	Opts() EngineOptions
+	// SetOpts sets the engine options.
+	SetOpts(EngineOptions) Engine
+}
+
+// EngineOptions can be used to pass custom flags to engine.
+type EngineOptions interface {
```

Comment:
Does this need to be an interface? Does it have multiple feasible implementations?

Reply:
I was just being consistent with what we do for other option types.

Reply:
Fair enough; I'm fine with leaving this. Ignore the rest of this comment for purposes

I'm going to write up a "Options interfaces considered harmful" doc one of these days--cc @prateek since I think you'll have an opinion on that  :). 

Briefly:

- It's lots of boilerplate
- There's only one implementation for all of these; mocking a plain-old-object/value struct makes little sense to me.
- It doesn't hide implementation details any better than just doing:

```
type MyOptions struct {
   // private
   f1 int
}

func (mo *MyOptions) F1() int {
}

// setters etc
```

I actually think we should do struct configuration more like the way most FX modules do, but it's definitely a broader conversation than this PR :)

Reply:
It's a fairly common pattern we use elsewhere in the codebase (e.g. [here](https://sourcegraph.com/github.com/m3db/m3@arnikola/multi-zone-coordinator/-/blob/src/dbnode/storage/types.go#L725:6))

Might not be super idiomatic go, but it's consistent with the codebase :+1:

Reply:
Oops, github hadn't updated with your comment haha, ignore me

Reply:
> Might not be super idiomatic go, but it's consistent with the codebase 👍

Hahaha I was going to comment with "oh trust me, I know". This is on my list of "Things I am (probably unnecessarily) salty about".